rdr_service/tools/tool_libs/backfill_coreminuspmtime.py

```diff
@@ -0,0 +1,99 @@
+import logging
+
+from rdr_service import clock
+from rdr_service.dao.participant_summary_dao import ParticipantSummaryDao
+from rdr_service.logic.enrollment_info import EnrollmentCalculation, EnrollmentDependencies
+from rdr_service.model.participant_summary import ParticipantSummary
+from rdr_service.participant_enums import EnrollmentStatus
+from rdr_service.repository.questionnaire_response_repository import QuestionnaireResponseRepository
+from rdr_service.services.system_utils import list_chunks, min_or_none
+from rdr_service.tools.tool_libs.tool_base import cli_run, ToolBase
+
+tool_cmd = 'backfill-coreminuspmtime'
+tool_desc = 'Backfill CoreMinusPMTime values that had been incorrectly updated'
+
+logger = logging.getLogger("rdr_logger")
+
+
+class BackfillCoreMinusPMTime(ToolBase):
+    def run(self):
+        super().run()
+        with self.get_session() as session:
+            participant_ids = session.query(
+                ParticipantSummary.participantId
+            ).filter(
+                ParticipantSummary.enrollmentStatusCoreMinusPMTime.isnot(None)
+            ).order_by(
+                ParticipantSummary.participantId
+            ).all()
+            summary_dao = ParticipantSummaryDao()
+            logger.info(f'PIDs to validate Core Minus PM Time: {len(participant_ids)}')
+            chunk_size = 50
+            count = 0
+            for participant_id_subset in list_chunks(lst=participant_ids, chunk_size=chunk_size):
+                batch_count = 0
+                logger.debug(f'pid subset: {participant_id_subset}')
+                logger.info(f'Validating PIDs {count*chunk_size} - {((count+1) * chunk_size)-1}')
+                summary_list = session.query(
+                    ParticipantSummary
+                ).filter(
+                    ParticipantSummary.participantId.in_(participant_id_subset)
+                ).with_for_update().all()
+                for summary in summary_list:
+                    calculated_date = self.calculate_core_minus_pm_time(summary, session)
+                    if calculated_date and calculated_date < summary.enrollmentStatusCoreMinusPMTime:
+                        batch_count += 1
+                        summary.enrollmentStatusCoreMinusPMTime = calculated_date
+                        summary.modifiedTime = clock.CLOCK.now()
+                        summary_dao.update(summary)
```

Comment:
This line can be removed (and might need to be considering how the database locking would work). You're loading the summaries further up and then setting the values on them. Later when you commit the session those changes will be sent to the database.

Reply:
I tested this out and that does conflict with the database lock so I've removed it now.